File: src/utils/timeline.ts
// ────────────────────────────────────────────────────────────────────────────
// Phase 6 — Réglette Temporelle Sémantique
// ────────────────────────────────────────────────────────────────────────────
// Module pur (testable, sans dépendance React/PixiJS).
//
// Convention : une "année" est un nombre entier signé.
//   • 1789  = an 1789 ap. J.-C.
//   • -3000 = 3000 av. J.-C.
//   • 0     = an 0 conventionnel (pas d'année 0 historique, mais on simplifie)
//   • -100000000 (-1e8) borne basse pratique (~Crétacé)
//   • 3000  borne haute pratique (futur proche)
//
// On évite Date / ms Unix : trop limité pour la préhistoire et négligeable
// pour la précision sub-annuelle dont on n'a pas besoin ici.

import type { TemporalAnchor } from "../types";

export const YEAR_MIN = -100_000_000; // -100 Ma (Crétacé moyen)
export const YEAR_MAX = 3_000;        // an 3000

// ── Époques nommées (extensibles par l'utilisateur plus tard) ─────────────
// Tableau ordonné chronologiquement. La première époque englobant l'année
// requise est retournée (les longues sont en premier, les précises ensuite).
export interface NamedEra {
  name: string;
  start: number;
  end: number;
  description?: string;
}

export const DEFAULT_ERAS: NamedEra[] = [
  // Préhistoire & paléontologie (large)
  { name: "Crétacé",            start: -145_000_000, end: -66_000_000, description: "Dinosaures, premiers oiseaux" },
  { name: "Paléogène",          start: -66_000_000,  end: -23_000_000, description: "Émergence des mammifères" },
  { name: "Néogène",            start: -23_000_000,  end: -2_580_000,  description: "Hominidés primitifs" },
  { name: "Pléistocène",        start: -2_580_000,   end: -11_700,     description: "Glaciations, Homo sapiens" },
  { name: "Holocène",           start: -11_700,      end: 2_026,       description: "Notre époque géologique" },

  // Préhistoire humaine
  { name: "Paléolithique",      start: -3_300_000,   end: -10_000,     description: "Pierre taillée" },
  { name: "Néolithique",        start: -10_000,      end: -3_300,      description: "Agriculture, sédentarité" },

  // Antiquité
  { name: "Antiquité",          start: -3_300,       end: 476,         description: "De l'écriture à la chute de Rome" },
  { name: "Égypte ancienne",    start: -3_150,       end: -30,         description: "Des premières dynasties à Cléopâtre" },
  { name: "Grèce antique",      start: -800,         end: -146,        description: "Cités-États, Alexandre" },
  { name: "République romaine", start: -509,         end: -27 },
  { name: "Empire romain",      start: -27,          end: 476 },

  // Moyen Âge
  { name: "Moyen Âge",          start: 476,          end: 1453,        description: "De la chute de Rome à celle de Constantinople" },
  { name: "Haut Moyen Âge",     start: 476,          end: 1000 },
  { name: "Moyen Âge central",  start: 1000,         end: 1300 },
  { name: "Bas Moyen Âge",      start: 1300,         end: 1453 },

  // Modernité
  { name: "Renaissance",        start: 1400,         end: 1600,        description: "Humanisme, redécouverte de l'antique" },
  { name: "Lumières",           start: 1715,         end: 1789,        description: "Raison, encyclopédie, droits naturels" },
  { name: "Révolution française", start: 1789,       end: 1799 },
  { name: "Empire napoléonien", start: 1804,         end: 1815 },
  { name: "Révolution industrielle", start: 1760,    end: 1840 },

  // XXe et après
  { name: "Belle Époque",       start: 1871,         end: 1914 },
  { name: "Première Guerre mondiale", start: 1914,   end: 1918 },
  { name: "Entre-deux-guerres", start: 1918,         end: 1939 },
  { name: "Seconde Guerre mondiale", start: 1939,    end: 1945 },
  { name: "Guerre froide",      start: 1947,         end: 1991 },
  { name: "Ère numérique",      start: 1990,         end: 2026,        description: "Web, mobile, IA générative" },
];

// ── Format d'une année ────────────────────────────────────────────────────
/**
 * Formate une année en string lisible selon son ordre de grandeur.
 *   1789      → "1789"
 *   -500      → "500 av. J.-C."
 *   -50000    → "50 ka"
 *   -1500000  → "1,5 Ma"
 */
export function formatYear(year: number): string {
  if (!Number.isFinite(year)) return "?";
  const y = Math.round(year);
  if (y >= 0) return String(y);
  const abs = -y;
  if (abs < 10_000) return `${abs} av. J.-C.`;
  if (abs < 1_000_000) return `${(abs / 1_000).toFixed(abs >= 100_000 ? 0 : 1)} ka`;
  return `${(abs / 1_000_000).toFixed(abs >= 100_000_000 ? 0 : 1)} Ma`;
}

/**
 * Formate une plage [start, end] de façon compacte.
 *   {1789, 1789} → "1789"
 *   {1789, 1799} → "1789–1799"
 */
export function formatAnchor(a: TemporalAnchor): string {
  if (a.label) return a.label;
  if (a.start === a.end) return formatYear(a.start);
  return `${formatYear(a.start)} – ${formatYear(a.end)}`;
}

// ── Parsing d'une saisie utilisateur ──────────────────────────────────────
/**
 * Convertit une saisie utilisateur en TemporalAnchor.
 * Formats acceptés :
 *   "1789"             → { start: 1789, end: 1789 }
 *   "1789-1799"        → { start: 1789, end: 1799 }
 *   "1789 - 1799"      → idem
 *   "-500"             → { start: -500, end: -500 }   (500 av. J.-C.)
 *   "500 av JC" / "500 BC" → { start: -500, end: -500 }
 *   "Renaissance"      → { start: 1400, end: 1600, label: "Renaissance" }
 *   "10 ka"            → { start: -10000, end: -10000 }
 *   "1,5 Ma"           → { start: -1500000, end: -1500000 }
 *
 * Renvoie null si non parsable.
 */
export function parseAnchor(input: string, eras: NamedEra[] = DEFAULT_ERAS): TemporalAnchor | null {
  const trimmed = input.trim();
  if (!trimmed) return null;

  // 1) Match exact d'une époque nommée (insensible à la casse)
  const lc = trimmed.toLowerCase();
  const era = eras.find((e) => e.name.toLowerCase() === lc);
  if (era) {
    return { start: era.start, end: era.end, label: era.name };
  }

  // 2) Plage "AAAA-BBBB" / "AAAA – BBBB" / "AAAA..BBBB"
  const range = trimmed.match(/^(-?\d[\d\s.,]*\s*(?:ka|ma|av\.?\s*j\.?-?\s*c\.?|bc)?)\s*(?:[–\-‒—]|\.\.)\s*(-?\d[\d\s.,]*\s*(?:ka|ma|av\.?\s*j\.?-?\s*c\.?|bc)?)$/i);
  if (range) {
    const a = parseSingleYear(range[1]);
    const b = parseSingleYear(range[2]);
    if (a !== null && b !== null) {
      return { start: Math.min(a, b), end: Math.max(a, b) };
    }
  }

  // 3) Année unique
  const single = parseSingleYear(trimmed);
  if (single !== null) {
    return { start: single, end: single };
  }

  return null;
}

/** Parse une année unique ; renvoie null si non reconnu. */
function parseSingleYear(input: string): number | null {
  const s = input.trim().toLowerCase().replace(/\s+/g, " ");

  // "1,5 ma", "1.5 Ma" → millions d'années avant maintenant (toujours négatif)
  const ma = s.match(/^(-?\d+(?:[.,]\d+)?)\s*ma$/);
  if (ma) {
    const n = Number(ma[1].replace(",", "."));
    if (Number.isFinite(n)) return -Math.round(n * 1_000_000);
  }

  // "10 ka", "12,5 ka" → milliers d'années avant maintenant
  const ka = s.match(/^(-?\d+(?:[.,]\d+)?)\s*ka$/);
  if (ka) {
    const n = Number(ka[1].replace(",", "."));
    if (Number.isFinite(n)) return -Math.round(n * 1_000);
  }

  // "500 av JC", "500 BC", "500 av. J.-C." → négatif
  const bc = s.match(/^(\d+)\s*(?:av\.?\s*j\.?-?\s*c\.?|bc)$/);
  if (bc) return -Number(bc[1]);

  // "1789 ap JC" / "1789 AD" → positif explicite
  const ad = s.match(/^(\d+)\s*(?:ap\.?\s*j\.?-?\s*c\.?|ad)$/);
  if (ad) return Number(ad[1]);

  // Entier signé brut "1789", "-500"
  const raw = s.match(/^(-?\d+)$/);
  if (raw) return Number(raw[1]);

  return null;
}

// ── Helper : un nœud passe-t-il le filtre temporel ? ──────────────────────
/**
 * Renvoie `true` si :
 *   - le nœud n'a pas de `temporalAnchor` (atemporel — toujours visible)
 *   - OU si l'intervalle du nœud chevauche celui du filtre (intersection non vide)
 */
export function nodeMatchesTemporalFilter(
  anchor: TemporalAnchor | undefined,
  filter: { start: number; end: number } | null
): boolean {
  if (!filter) return true;
  if (!anchor) return true;
  return anchor.start <= filter.end && anchor.end >= filter.start;
}

// ── Échelles de zoom de la réglette ───────────────────────────────────────
/** Retourne le pas de graduation adapté à la largeur visible. */
export function tickStep(spanYears: number): number {
  if (spanYears > 50_000_000) return 10_000_000;
  if (spanYears > 5_000_000)  return 1_000_000;
  if (spanYears > 500_000)    return 100_000;
  if (spanYears > 50_000)     return 10_000;
  if (spanYears > 5_000)      return 1_000;
  if (spanYears > 500)        return 100;
  if (spanYears > 50)         return 10;
  if (spanYears > 5)          return 1;
  return 1;
}

File: crates/glucose-core/src/timeline.rs
//! Phase 6 — Réglette Temporelle Sémantique (0 dépendance).

use crate::types::TemporalAnchor;

pub const YEAR_MIN: i64 = -100_000_000;
pub const YEAR_MAX: i64 = 3_000;

#[derive(Debug, Clone, PartialEq)]
pub struct NamedEra {
    pub name: &'static str,
    pub start: i64,
    pub end: i64,
    pub description: Option<&'static str>,
}

pub const DEFAULT_ERAS: &[NamedEra] = &[
    NamedEra { name: "Crétacé", start: -145_000_000, end: -66_000_000, description: Some("Dinosaures, premiers oiseaux") },
    NamedEra { name: "Paléogène", start: -66_000_000, end: -23_000_000, description: Some("Émergence des mammifères") },
    NamedEra { name: "Néogène", start: -23_000_000, end: -2_580_000, description: Some("Hominidés primitifs") },
    NamedEra { name: "Pléistocène", start: -2_580_000, end: -11_700, description: Some("Glaciations, Homo sapiens") },
    NamedEra { name: "Holocène", start: -11_700, end: 2026, description: Some("Notre époque géologique") },

    NamedEra { name: "Paléolithique", start: -3_300_000, end: -10_000, description: Some("Pierre taillée") },
    NamedEra { name: "Néolithique", start: -10_000, end: -3_300, description: Some("Agriculture, sédentarité") },

    NamedEra { name: "Antiquité", start: -3_300, end: 476, description: Some("De l'écriture à la chute de Rome") },
    NamedEra { name: "Égypte ancienne", start: -3_150, end: -30, description: Some("Des premières dynasties à Cléopâtre") },
    NamedEra { name: "Grèce antique", start: -800, end: -146, description: Some("Cités-États, Alexandre") },
    NamedEra { name: "République romaine", start: -509, end: -27, description: None },
    NamedEra { name: "Empire romain", start: -27, end: 476, description: None },

    NamedEra { name: "Moyen Âge", start: 476, end: 1453, description: Some("De la chute de Rome à celle de Constantinople") },
    NamedEra { name: "Haut Moyen Âge", start: 476, end: 1000, description: None },
    NamedEra { name: "Moyen Âge central", start: 1000, end: 1300, description: None },
    NamedEra { name: "Bas Moyen Âge", start: 1300, end: 1453, description: None },

    NamedEra { name: "Renaissance", start: 1400, end: 1600, description: Some("Humanisme, redécouverte de l'antique") },
    NamedEra { name: "Lumières", start: 1715, end: 1789, description: Some("Raison, encyclopédie, droits naturels") },
    NamedEra { name: "Révolution française", start: 1789, end: 1799, description: None },
    NamedEra { name: "Empire napoléonien", start: 1804, end: 1815, description: None },
    NamedEra { name: "Révolution industrielle", start: 1760, end: 1840, description: None },

    NamedEra { name: "Belle Époque", start: 1871, end: 1914, description: None },
    NamedEra { name: "Première Guerre mondiale", start: 1914, end: 1918, description: None },
    NamedEra { name: "Entre-deux-guerres", start: 1918, end: 1939, description: None },
    NamedEra { name: "Seconde Guerre mondiale", start: 1939, end: 1945, description: None },
    NamedEra { name: "Guerre froide", start: 1947, end: 1991, description: None },
    NamedEra { name: "Ère numérique", start: 1990, end: 2026, description: Some("Web, mobile, IA générative") },
];

pub fn clamp_year(year: i64) -> i64 {
    year.clamp(YEAR_MIN, YEAR_MAX)
}

pub fn format_year(year: i64) -> String {
    if year >= 0 {
        return year.to_string();
    }
    let abs = -year;
    if abs < 10_000 {
        format!("{} av. J.-C.", abs)
    } else if abs < 1_000_000 {
        let ka = abs as f64 / 1_000.0;
        if abs >= 100_000 {
            format!("{:.0} ka", ka)
        } else {
            format!("{:.1} ka", ka)
        }
    } else {
        let ma = abs as f64 / 1_000_000.0;
        if abs >= 100_000_000 {
            format!("{:.0} Ma", ma)
        } else {
            format!("{:.1} Ma", ma)
        }
    }
}

pub fn format_anchor(a: &TemporalAnchor) -> String {
    if let Some(ref label) = a.label {
        return label.clone();
    }
    if a.start == a.end {
        format_year(a.start)
    } else {
        format!("{} – {}", format_year(a.start), format_year(a.end))
    }
}

fn parse_single_year(input: &str) -> Option<i64> {
    let s = input.trim().to_lowercase();
    let s = s.split_whitespace().collect::<Vec<&str>>().join(" ");

    if s.ends_with("ma") {
        let num_part = s[..s.len() - 2].trim().replace(',', ".");
        if let Ok(val) = num_part.parse::<f64>() {
            return Some(-(val * 1_000_000.0).round() as i64);
        }
    }

    if s.ends_with("ka") {
        let num_part = s[..s.len() - 2].trim().replace(',', ".");
        if let Ok(val) = num_part.parse::<f64>() {
            return Some(-(val * 1_000.0).round() as i64);
        }
    }

    let bc_suffixes = [
        "av. j.-c.", "av. j-c", "av jc", "av. jc", "av j.-c.", "av. j.-c", "bc",
    ];
    for suf in &bc_suffixes {
        if s.ends_with(suf) {
            let num_part = s[..s.len() - suf.len()].trim();
            if let Ok(val) = num_part.parse::<i64>() {
                return Some(-val);
            }
        }
    }

    let ad_suffixes = [
        "ap. j.-c.", "ap. j-c", "ap jc", "ap. jc", "ap j.-c.", "ad",
    ];
    for suf in &ad_suffixes {
        if s.ends_with(suf) {
            let num_part = s[..s.len() - suf.len()].trim();
            if let Ok(val) = num_part.parse::<i64>() {
                return Some(val);
            }
        }
    }

    if let Ok(val) = s.parse::<i64>() {
        return Some(val);
    }

    None
}

pub fn parse_anchor(input: &str, eras: &[NamedEra]) -> Option<TemporalAnchor> {
    let trimmed = input.trim();
    if trimmed.is_empty() {
        return None;
    }

    let lc = trimmed.to_lowercase();
    if let Some(era) = eras.iter().find(|e| e.name.to_lowercase() == lc) {
        return Some(TemporalAnchor {
            start: era.start,
            end: era.end,
            label: Some(era.name.to_string()),
        });
    }

    let range_seps = [" – ", " - ", "—", "‒", "..", "–", "-"];
    for sep in &range_seps {
        let search_start = if trimmed.starts_with('-') && (*sep == "-" || *sep == "–") { 1 } else { 0 };
        if let Some(pos) = trimmed[search_start..].find(sep) {
            let abs_pos = search_start + pos;
            let left = &trimmed[..abs_pos];
            let right = &trimmed[abs_pos + sep.len()..];
            if !left.is_empty() && !right.is_empty() {
                if let (Some(a), Some(b)) = (parse_single_year(left), parse_single_year(right)) {
                    return Some(TemporalAnchor {
                        start: a.min(b),
                        end: a.max(b),
                        label: None,
                    });
                }
            }
        }
    }

    if let Some(y) = parse_single_year(trimmed) {
        return Some(TemporalAnchor {
            start: y,
            end: y,
            label: None,
        });
    }

    None
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TemporalFilter {
    pub start: i64,
    pub end: i64,
}

pub fn node_matches_temporal_filter(
    anchor: Option<&TemporalAnchor>,
    filter: Option<TemporalFilter>,
) -> bool {
    let f = match filter {
        Some(f) => f,
        None => return true,
    };
    let a = match anchor {
        Some(a) => a,
        None => return true,
    };
    a.start <= f.end && a.end >= f.start
}

pub fn tick_step(span_years: i64) -> i64 {
    if span_years > 50_000_000 {
        10_000_000
    } else if span_years > 5_000_000 {
        1_000_000
    } else if span_years > 500_000 {
        100_000
    } else if span_years > 50_000 {
        10_000
    } else if span_years > 5_000 {
        1_000
    } else if span_years > 500 {
        100
    } else if span_years > 50 {
        10
    } else {
        1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_year() {
        assert_eq!(format_year(1789), "1789");
        assert_eq!(format_year(-500), "500 av. J.-C.");
        assert_eq!(format_year(-50_000), "50.0 ka");
        assert_eq!(format_year(-1_500_000), "1.5 Ma");
    }

    #[test]
    fn test_format_anchor() {
        let single = TemporalAnchor { start: 1789, end: 1789, label: None };
        assert_eq!(format_anchor(&single), "1789");

        let span = TemporalAnchor { start: 1789, end: 1799, label: None };
        assert_eq!(format_anchor(&span), "1789 – 1799");

        let labeled = TemporalAnchor { start: 1789, end: 1799, label: Some("Révolution".into()) };
        assert_eq!(format_anchor(&labeled), "Révolution");
    }

    #[test]
    fn test_parse_anchor_era() {
        let anchor = parse_anchor("Renaissance", DEFAULT_ERAS).expect("Renaissance");
        assert_eq!(anchor.start, 1400);
        assert_eq!(anchor.end, 1600);
        assert_eq!(anchor.label.as_deref(), Some("Renaissance"));
    }
}
